enterprise_access/utils.py

```diff
@@ -285,3 +289,81 @@ def get_course_run_metadata_for_assignment(assignment, content_metadata):
 
     # For course-based assignments, return metadata for the advertised course run
     return get_advertised_course_run_metadata(content_metadata)
+
+
+def time_until(date_str: str, unit: str = 'seconds', exact: bool = False) -> Union[float, int]:
+    """
+    Calculates the time difference between the current time and a given ISO 8601 datetime string.
+
+    The input string must be timezone-aware (e.g., include 'Z' for UTC or a timezone offset like '+00:00').
+    The result is returned in the specified time unit (case-insensitive), either as a float or an integer,
+    depending on the `exact` flag. The result may be negative if the given time is in the past.
+
+    Args:
+        date_str (str): ISO 8601 datetime string (e.g., '2025-03-25T17:45:05Z').
+                        Must include timezone information.
+        unit (str): Unit for the result — 'seconds', 'minutes', 'hours', or 'days'. Default is 'seconds'.
+        exact (bool): If True, returns a float (e.g., 91.42 seconds).
+                      If False, returns the floored integer value (e.g., 91).
+
+    Returns:
+        float | int: Time difference in the specified unit. May be negative if the time is in the past.
+
+    Raises:
+        ValueError: If the date string is invalid, lacks timezone information,
+                    or if an unsupported unit is specified.
+    """
+    parsed_time: Optional[datetime] = dateparse.parse_datetime(date_str)
+    if not parsed_time or not timezone.is_aware(parsed_time):
+        raise ValueError(f"Invalid or non-timezone-aware date string: {date_str!r}")
+
+    # Calculate the difference in seconds
+    time_delta_seconds: float = (parsed_time - timezone.now()).total_seconds()
+
+    # Convert based on unit
+    match unit.lower():
+        case 'seconds':
+            result = time_delta_seconds
+        case 'minutes':
+            result = time_delta_seconds / 60
+        case 'hours':
+            result = time_delta_seconds / 3600
+        case 'days':
+            result = time_delta_seconds / 86400
+        case _:
+            raise ValueError("Invalid unit. Use 'seconds', 'minutes', or 'days'.")
+
+    return result if exact else math.floor(result)
+
+
+def determine_timeout_offset(
+        valid_until_date_str: str,
+        epsilon_seconds: int = 10,
+        timeout: int = DEFAULT_CACHE_TIMEOUT
+) -> Union[float, int]:
+    """
+    Determines the effective timeout duration based on a 'valid until' UTC timestamp,
+    a fallback timeout in seconds, and an epsilon buffer.
+
+    This function uses `time_until` to calculate the time (in seconds) until the provided
+    `valid_until_date_str`. If the result is negative (i.e., the time is in the past), the
+    fallback `timeout` value is used instead.
```

Comment:
I'm not quite following why "[i]f the result is negative (i.e., the time is in the past), the fallback `timeout` value is used instead."

Can you help me understand why if the `valid_until_date_str` is already in the past, we'd want to fallback to the `DEFAULT_CACHE_TIMEOUT` of 5 minutes? If it's already in the past (which shouldn't really happen in practice for this specific use case), I think we'd want the timeout to be `0` to force a re-fetch to get a new `valid_until`? 🤔